Initial Implementation of Basic Auth

richteer opened the pull request:
FYI: Don't merge me :)  Still depends on unique identifiers. Also, I haven't actually tested on live yet, just with the test cases. Close enough, right?

See commit messages for explanantions.

richteer committed:
halauth: add authentication object

HalAuth is a new object intended to sit in Halibot core, and handles permission
objects.

Authentication objects currently take the form of a triple:
(RI, identity, Permission String)
...which binds a particular permission to a identity and RI.

An example would be the following:
(irc/##halibot, richteer, CanReloadModules)
...which essentially says that the user "richteer" in the IRC channel "##halibot"
is allowed to operate anything that requests the "CanReloadModules" permission.

Any of the three can be wildcarded, including all of them at once. Currently,
the wildcard indicator is an asterisk ("*"). This indicates that ANY of that
particular field matches, so allowing all permissions to a particular user in
a particular channel (RI) would look like this:
(*, richteer, *)

The current API on the HalAuth module includes three methods:
 - grantPermission(ri, identity, perm) -> add a perm triple to the list
 - revokePermission(ri, identity, perm) -> remove an EXACT MATCHING triple
 - hasPermission(ri, identity, perm) -> check if a triple matches one in the list

richteer committed:
halibot: integrate halauth into halibot core

Add the HalAuth object into Halibot core.

An optional config key has been added: "auth-path". This points to a particular
file which holds the permissions list. If not supplied, the default path of
"$CURRENTDIR/permissions.json" is used.

HalAuth can be disabled either by setting the "use-auth" key to False.Disabling
auth puts the HalAuth object in a permissive mode, that always returns True.
This is currently for legacy compatibility, but should probably be configurable.

richteer committed:
halmodule: add hasPermission helper func for module developers

Add a simple wrapper to make the call to HalAuth's hasPermission().

This implementation takes in a message object for simplicity.

richteer committed:
tests: add initial halauth testing, only hasPermission methods for now

richteer committed:
halauth: add @hasPermission() module method decorator

Implements a function decorator for easy permission checking at
the method-level.

The decorater takes in a permission string as an argument, and expects
to be attached to a method with a message object as the first argument.

Example module usage: (and main use-case):

```
class ExampleModule

	@hasPermission("CanCommand")
	def SomeCommand(self, msg):
		# protected code here...

	def receive(self, msg):
		if msg.body.startswith("!command"):
			self.SomeCommand(msg)
```

richteer commented on halibot/halmodule.py:
```diff
@@ -17,3 +17,5 @@ def reply(self, msg0=None, **kwargs):
 		self.send_to(msg, [ msg.origin ])
 
+	def hasPermission(self, msg, perm):
+		return self._hal.auth.hasPermission(msg.origin, msg.author, perm)
```
This should use `msg.author.id`, when that is implemented.

richteer commented on halibot/halauth.py:
```diff
@@ -1,6 +1,16 @@
+def hasPermission(perm, reply=False):
+	def real_dec(func):
+		def wrapper(self, msg, *args, **kwargs):
+			if self._hal.auth.hasPermission(msg.origin, msg.author, perm):
```
Same as the HalMessage method, needs `msg.author.id`

richteer commented on halibot/halauth.py:
```diff
@@ -1,6 +1,16 @@
 import json
 import logging
 
+def hasPermission(perm, reply=False):
```
Maybe make the auto-reply a configuration option?

If so, where, main config, or per module/instance?

I can almost see a case where some channels may want to silence all failures, but others may still want them.

sjrct commented on halibot/halauth.py:
```diff
@@ -0,0 +1,77 @@
+			if self._hal.auth.hasPermission(msg.origin, msg.author, perm):
+				func(self, msg, *args, **kwargs)
+			elif reply:
+				self.reply(msg, body="Permission Denied")
```
This message should probably be configurable as well.

richteer replied:
Thoughts on `self.config.get("no-perm","Permission Denied")` ? Probably use a better config key than `no-perm`, but it would be an optional key on all modules.

Alternatively, having a halibot-global option.

Also, the `reply` bool should probably be a config option, not at the annotation level

sjrct replied:
Something like that should be fine. Something like `perm-denied-msg` might be better, but I am not too picky. When it comes to config v. annotation argument for this and for the `reply` bool, could have them in the config and also have an optional argument that when set overrides the config setting.

sjrct commented on halibot/halauth.py:
```diff
@@ -0,0 +1,77 @@
+
+		except Exception as e:
+			self.log.error("Error loading permissions: {}".format(e))
+			# Return if can't find auth?
```
I think we can safely assume that no permissions file should behave the same as an empty one.

sjrct commented on halibot/halibot.py:
```diff
@@ -28,6 +29,9 @@ def __init__(self, **kwargs):
 		self.use_config = kwargs.get("use_config", True)
+		self.use_auth = kwargs.get("use_auth", True)
+
+		self.auth = HalAuth()
```
Should the value of `use_auth` set the value of `auth.enabled`? It seems a little redundant to have bolth.

richteer replied:
Uhhhh probably. Will take a second look at. I added `auth.enabled` after `use_auth`, and forgot about the latter tbh :)

sjrct commented on halibot/halauth.py:
```diff
@@ -0,0 +1,77 @@
+		self.path = path
+
+		try:
+			with open(self.path, "r") as f:
```
Should we consider placing the permissions inside the regular top-level `config.json`? What is the rational for having a separate file?

richteer replied:
Permissions will potentially get very large, keeps from cluttering the main one. Also, I'm not 100% sure that permissions will stay JSON, since the data doesn't really lend itself to a key-value form well. (More of a value-value-value).

sjrct replied:
Fair enough. These are probably more likely to be auto-generated than the normal config.json as well.

sjrct reviewed (changes requested):
Looks pretty good to me. The only thing I would say is that this does not really become usable until we provide the ability to set the `id` field on the `author`. I don't think it would be too hard to tack that onto this PR, and it would provide the minimum needed for 1.0 for auth in the core.

richteer commented:
Yep, will open a new PR (after oct 1st...) for the `author.id` changes, will look to merge that one first.

sjrct commented:
As long as the author changes are incoming this, it is fine. It is just that these changes are not practically usable without the author changes.

sjrct commented:
Should `hasPermission` be on `HalObject` instead of `HalModule`, making it available to both modules and agents?

richteer committed:
message: add identity field to messages

Adds a field for the author's uniquely identifying string, for purposes such as
permissions, and per-user data storage.
